### DISCUSSION TITLE
WGD HA setup - Queries and Ideas.

### DISCUSSION THREAD

**Discussion by mzs114** [WGDashboard/WGDashboard#610] [Ideas]
**Is your feature request related to a problem? Please describe.**
It seems we have have a HA setup with some cloud like AWS, and WGD can support this if the state(DB) sharing is taken care of.

Refer the below setup:
https://www.procustodibus.com/blog/2021/02/ha-wireguard-on-aws/

**Describe the solution you'd like**

Can WGD support PostgreSQL, so that one can setup AWS Aurora for shared state between multiple WGD VMs?

**Comment by DaanSelen** [comment]
You mean support an SQLike database such as Postgres or MySQL/MariaDB?

**Comment by donaldzou** [comment]
Theoretically... it can be done.. since the SQL statement I used works across different flavours of SQL DBs. One thing I would like to know is, if I understand HA WireGuard setup correctly: 
1. It uses multiple servers but assign them by load balancer
2. You will need identical WG configurations across servers
3. Deploy WGDashboard on each servers
    - If you use **one** database across all WGDashboard instances, some records in databse will get override such as:
        - ServerA => Peer1 recorded using 4.4GB total
        - ServerB => Peer1 recorded using 1GB total
        - **In this case, you will see different data usage whenever each server update**

Correct me if I'm wrong. But using PostgreSQL, MySQL, SQL Server is possible with WGDashboard.

**Comment by mzs114** [comment]
> You mean support an SQLike database such as Postgres or MySQL/MariaDB?

Yes, any other FLOSS RDBMS, that supports external HA setup.



> Theoretically... it can be done.. since the SQL statement I used works across different flavours of SQL DBs. One thing I would like to know is, if I understand HA WireGuard setup correctly:
> 

An ORM may help, but not sure how Flask handles this.


>     1. It uses multiple servers but assign them by load balancer
> 

Yes.

>     2. You will need identical WG configurations across servers
> 

Correct.

>     3. Deploy WGDashboard on each servers
>        

Correct, and they all connect to the same DB.

>        * If you use **one** database across all WGDashboard instances, some records in databse will get override such as:
>          
>          * ServerA => Peer1 recorded using 4.4GB total
>          * ServerB => Peer1 recorded using 1GB total
>          * **In this case, you will see different data usage whenever each server update**
> 
> 

This could be a problem, but I think if we use sticky sessions the connection will be routed to the same backend node, however, this is still not a perfect setup, as the peer can switch after idle timeout to a new backend node behind the load balancer, need to see how other FLOSS products are handling this.

https://cloudar.be/awsblog/why-aws-nlb-stickiness-is-not-always-sticky/

**Comment by donaldzou** [comment]
> This could be a problem, but I think if we use sticky sessions the connection will be routed to the same backend node, however, this is still not a perfect setup, as the peer can switch after idle timeout to a new backend node behind the load balancer, need to see how other FLOSS products are handling this.

I see, but I'm thinking if we use sticky session, isn't it kinda defeated the purpose of using NLB lol.. Currently WGDashboard support cross server access on its Desktop app, maybe this would be a better solution?

**Comment by mzs114** [comment]
> I see, but I'm thinking if we use sticky session, isn't it kinda defeated the purpose of using NLB lol.. Currently WGDashboard support cross server access on its Desktop app, maybe this would be a better solution?

Some Wireguard uses UDP, even if the back node is lost, the connection routes through some other node, ideally the user would not notice.

**Comment by donaldzou** [comment]
I will convert this to a discussion :)
